File: solution/19. Remove Nth Node From End of List/19.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode dummy(0);
        dummy.next = head;
        
        auto fast = &dummy;
        for (int i = 0; i < n; ++i) {
            fast = fast->next;
        }
        
        auto slow = &dummy;
        while (fast->next) {
            slow = slow->next;
            fast = fast->next;
        }
        
        auto tmp = slow->next;
        slow->next = slow->next->next;
        delete tmp;
        
        return dummy.next;
    }
};

File: solution/19. Remove Nth Node From End of List/19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        
        tail = dummy
        while n > 0:
            tail = tail.next
            n -= 1
        
        pre = dummy
        while tail.next:
            pre = pre.next
            tail = tail.next
        
        pre.next = pre.next.next
        
        return dummy.next
    
